From aef8525b69a7b4ed2aab62f69354e6b7b8126112 Mon Sep 17 00:00:00 2001
From: Wolfgang Grandegger <wg@aries-embedded.de>
Date: Sat, 26 Feb 2022 13:02:51 +0100
Subject: [PATCH 7/7] ARM: socfpga: gen5: release reset before using USB as
 host with usb start

The command "usb start" does not work because the USB device is still in
reset. Releasing it in board_usb_init() fixes the problem. This issue
has been observed on the MCVEVK board.

Signed-off-by: Wolfgang Grandegger <wg@aries-embedded.de>
---
 arch/arm/mach-socfpga/board.c | 14 ++++++++++----
 drivers/usb/host/dwc2.c       | 12 ++++++++++++
 2 files changed, 22 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mach-socfpga/board.c b/arch/arm/mach-socfpga/board.c
index be4ad29d9b..03f9e3076a 100644
--- a/arch/arm/mach-socfpga/board.c
+++ b/arch/arm/mach-socfpga/board.c
@@ -57,7 +57,9 @@ int dram_init_banksize(void)
 struct dwc2_plat_otg_data socfpga_otg_data = {
 	.usb_gusbcfg	= 0x1417,
 };
+#endif
 
+#if defined(CONFIG_USB_GADGET) || defined(CONFIG_USB_DM)
 int board_usb_init(int index, enum usb_init_type init)
 {
 	int node[2], count;
@@ -75,9 +77,6 @@ int board_usb_init(int index, enum usb_init_type init)
 		return -EINVAL;
 	}
 
-	/* Patch the address from OF into the controller pdata. */
-	socfpga_otg_data.regs_otg = addr;
-
 #ifdef CONFIG_TARGET_SOCFPGA_GEN5
 	/* First release reset of the USB port */
 	if (addr == SOCFPGA_USB0_ADDRESS)
@@ -86,7 +85,14 @@ int board_usb_init(int index, enum usb_init_type init)
 		socfpga_per_reset(SOCFPGA_RESET(USB1), 0);
 #endif
 
-	return dwc2_udc_probe(&socfpga_otg_data);
+#ifdef CONFIG_USB_GADGET
+	if (init == USB_INIT_DEVICE) {
+		/* Patch the address from OF into the controller pdata. */
+		socfpga_otg_data.regs_otg = addr;
+		return dwc2_udc_probe(&socfpga_otg_data);
+	}
+#endif
+	return 0;
 }
 
 int g_dnl_board_usb_cable_connected(void)
diff --git a/drivers/usb/host/dwc2.c b/drivers/usb/host/dwc2.c
index b9c56f763b..0e49e8d37a 100644
--- a/drivers/usb/host/dwc2.c
+++ b/drivers/usb/host/dwc2.c
@@ -1321,13 +1321,25 @@ static int dwc2_usb_ofdata_to_platdata(struct udevice *dev)
 	return 0;
 }
 
+
+__weak int __board_usb_init(int index, enum usb_init_type init)
+{
+	return 0;
+}
+
 static int dwc2_usb_probe(struct udevice *dev)
 {
 	struct dwc2_priv *priv = dev_get_priv(dev);
 	struct usb_bus_priv *bus_priv = dev_get_uclass_priv(dev);
+	int ret;
 
 	bus_priv->desc_before_addr = true;
 
+	/* board-dependant init */
+	ret = board_usb_init(dev->seq, USB_INIT_HOST);
+	if (ret)
+		return ret;
+
 	return dwc2_init_common(dev, priv);
 }
 
-- 
2.28.0

